File: source/RolXServer/RolXServer/Projects/Domain/Model/Subproject.cs
// -----------------------------------------------------------------------
// <copyright file="Subproject.cs" company="Christian Ewald">
// Copyright (c) Christian Ewald. All rights reserved.
// Licensed under the MIT license.
// See LICENSE.md in the project root for full license information.
// </copyright>
// -----------------------------------------------------------------------

using RolXServer.Common.Util;
using RolXServer.Users.DataAccess;

namespace RolXServer.Projects.Domain.Model;

/// <summary>
/// A subproject we are working on.
/// </summary>
public sealed class Subproject
{
    /// <summary>
    /// Gets or sets the identifier.
    /// </summary>
    public int Id { get; set; }

    /// <summary>
    /// Gets or sets the number.
    /// </summary>
    public int Number { get; set; }

    /// <summary>
    /// Gets or sets the name.
    /// </summary>
    public string Name { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the project number.
    /// </summary>
    public int ProjectNumber { get; set; }

    /// <summary>
    /// Gets the full-qualified number of the specified subproject.
    /// </summary>
    public string FullNumber
    {
        get { return $"#{this.ProjectNumber:D4}.{this.Number:D3}"; }
    }

    /// <summary>
    /// Gets or sets the project name.
    /// </summary>
    public string ProjectName { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the name of the customer.
    /// </summary>
    public string CustomerName { get; set; } = string.Empty;

    /// <summary>
    /// Gets all names of the specified subproject.
    /// </summary>
    public string AllNames
    {
        get { return $"{this.CustomerName} - {this.ProjectName} - {this.Name}"; }
    }

    /// <summary>
    /// Gets the full-qualified name of the specified subproject.
    /// </summary>
    public string FullName
    {
        get { return $"{this.AllNames} ({this.FullNumber})"; }
    }

    /// <summary>
    /// Gets or sets the activities.
    /// </summary>
    public List<Activity> Activities { get; set; } = new List<Activity>();

    /// <summary>
    /// Gets a value indicating whether this instance is closed.
    /// </summary>
    public bool IsClosed
    {
        get
        {
            return this.Activities
                .All(activity => activity.EndedDate != null && activity.EndedDate <= DateOnly.FromDateTime(DateTime.Now));
        }
    }

    /// <summary>
    /// Gets or sets the manager identifier.
    /// </summary>
    public Guid? ManagerId { get; set; }

    /// <summary>
    /// Gets or sets the manager.
    /// </summary>
    public User? Manager { get; set; }

    /// <summary>
    /// Gets the budget time.
    /// </summary>
    public TimeSpan Budget => this.Activities
        .Where(a => a.Budget.HasValue)
        .Sum(a => a.Budget!.Value);

    /// <summary>
    /// Gets the planned time.
    /// </summary>
    public TimeSpan Planned => this.Activities
        .Where(a => a.Planned.HasValue)
        .Sum(a => a.Planned!.Value);

    /// <summary>
    /// Gets the actual time.
    /// </summary>
    public TimeSpan Actual => this.Activities
        .Where(a => a.Actual.HasValue)
        .Sum(a => a.Actual!.Value);

    /// <summary>
    /// Gets the ratio of actual time to budgeted time.
    /// </summary>
    public double? BudgetConsumedFraction
        => this.Activities.Any(a => a.Actual.HasValue) && this.Activities.Any(a => a.Budget.HasValue)
        ? this.Actual.TotalSeconds / this.Budget.TotalSeconds
        : null;

    /// <summary>
    /// Gets the ratio of actual time to planned time.
    /// </summary>
    public double? PlannedConsumedFraction
        => this.Activities.Any(a => a.Actual.HasValue) && this.Activities.Any(a => a.Planned.HasValue)
        ? this.Actual.TotalSeconds / this.Planned.TotalSeconds
        : null;

    /// <summary>
    /// Gets a value indicating whether the actual time is larger than the time budget.
    /// </summary>
    public bool IsOverBudget
        => this.Activities.Any(a => a.Budget.HasValue)
        && this.Actual > this.Budget;

    /// <summary>
    /// Gets a value indicating whether the actual time is larger than the planned time.
    /// </summary>
    public bool IsOverPlanned
        => this.Activities.Any(a => a.Planned.HasValue)
        && this.Actual > this.Planned;
}

File: source/rolx-client/src/app/projects/core/subproject.ts
import { Duration, TransformAsDuration } from '@app/core/util/duration';
import { TransformAsIsoDate } from '@app/core/util/iso-date';
import { assertDefined } from '@app/core/util/utils';
import { SubprojectShallow } from '@app/projects/core/subproject-shallow';
import { Type } from 'class-transformer';
import * as moment from 'moment';

import { Activity } from './activity';

export class Subproject extends SubprojectShallow {
  fullName!: string;
  projectNumber!: number;
  number!: number;
  managerId?: string;
  deputyManagerId?: string;
  architectId?: string;
  isOverBudget!: boolean;
  isOverPlanned!: boolean;

  @TransformAsIsoDate()
  startDate!: moment.Moment | null;

  @TransformAsDuration()
  budget!: Duration;

  @TransformAsDuration()
  planned!: Duration;

  @TransformAsDuration()
  actual!: Duration;

  budgetConsumedFraction!: number;
  plannedConsumedFraction!: number;

  @Type(() => Activity)
  activities: Activity[] = [];

  override validateModel(): void {
    super.validateModel();

    assertDefined(this, 'fullName');
    assertDefined(this, 'projectNumber');
    assertDefined(this, 'number');
    assertDefined(this, 'isOverBudget');
    assertDefined(this, 'budget');
    assertDefined(this, 'planned');
    assertDefined(this, 'actual');
    assertDefined(this, 'budgetConsumedFraction');
    assertDefined(this, 'plannedConsumedFraction');
    assertDefined(this, 'activities');

    this.activities.forEach((a) => a.validateModel());
  }

  get nextNumber() {
    return Math.max(...this.activities.map((ph) => ph.number), 0) + 1;
  }

  addActivity(): Activity {
    const activity = new Activity();
    activity.startDate = moment();
    activity.number = this.nextNumber;
    activity.fullName = '';
    activity.fullNumber = '';
    activity.allSubprojectNames = '';
    activity.billabilityName = '';
    activity.customerName = '';
    activity.projectName = '';
    activity.subprojectName = '';
    this.activities.push(activity);
    return activity;
  }
}
